[DOC-430] Document Kafka support (#309)
Macrometacorp/docs: docs/streams/kafka_on_gdn/kafkajs-client.md

---
sidebar_position: 40
title: KafkaJS Client
---

The KafkaJS client can be used to create Kafka producers, consumers. Full documentation [here](https://kafka.js.org/docs/getting-started).

### Installation

> Install KafkaJS using:

Reviewer: Do we need to tell people how to install or can you just link to the documentation?

Author: This step would be required as broker names are different in our case

```shell
yarn add kafkajs
```

or npm:

```shell
npm install kafkajs
```

> Let's start by instantiating the KafkaJS client by pointing it towards at least one broker:

Reviewer: ```suggestion
Instantiate the KafkaJS client by pointing it towards at least one broker:
```

Reviewer: I'm confused. Previous topics read like reference, this reads more like a conversational tutorial. We should keep the tone consistent.

```node
const { Kafka } = require('kafkajs')

const kafka = new Kafka({
  clientId: '<my-app>',
  brokers: ['<my-gdn>.<my-paas>.macrometa.io:9093'],
  sasl: { mechanism: 'plain', username: "<my-tenant>/<my-fabric>", password: "token:<my-api-key>" },
    ssl: {
    rejectUnauthorized: false
    }
})
```

### Producer

> To produce a message to a topic, we'll create a producer using our client:

Reviewer: ```suggestion
To produce a message to a topic, create a producer using our client:
```

Reviewer: What is "our client"? Macrometa? The Kafka client? Unclear.

```node
const producer = kafka.producer()

await producer.connect()
await producer.send({
  topic: '<c8globals or c8locals>.<my-topic>',
  messages: [
    { value: 'Hello KafkaJS user!' },
  ],
})

await producer.disconnect()
```

### Consumer

> To verify that our message has indeed been produced to the topic, let's create a consumer to consume our message:

Reviewer: ```suggestion
To verify the message has been sent to the topic, create a consumer to consume the message:
```

```node
const consumer = kafka.consumer({ groupId: '<my-group-id>' })

await consumer.connect()

> await consumer.subscribe({ topic: '<c8globals or c8locals>.<my-topic>', fromBeginning: true })

Reviewer: Make sure you remove all mention of "c8" throughout.
